common/event_emitter.go

```diff
@@ -180,6 +187,17 @@ func (e *BaseEventEmitter) emit(event string, data interface{}) {
 				handlers = append(handlers[:i], handlers[i+1:]...)
 				continue
 			default:
+				// To ensure that goroutines don't break the ordering
+				// of the emitted events, we will need the goroutine to synchronize.
+				// This means that the events need to be stored in a queue per handler. The
+				// goroutine responsible for the the first popped element must complete
+				// publishing the event before the next goroutine can pop and work with
+				// the next event for that one handler. Each handler can process events concurrently still.
+
+				handler.queueMutex.Lock()
+				handler.queue = append(handler.queue, Event{typ: event, data: data})
+				handler.queueMutex.Unlock()
```

Comment:
A channel would accomplish the same thing, no? It's a FIFO queue after all, and we wouldn't need locking.

Reply:
A channel can't grow as the complexity of the tests increase and thus the number of events increase. So i think the queue is a better approach for this case.

Reply:
I tried that too! 😆 I wish I had saved it as a PR 😞 I'll recreate it and create a PR for you to look at.

Reply:
Maybe something like this?

<details>
  <summary>Channel approach</summary>
  
```diff
diff --git a/common/event_emitter.go b/common/event_emitter.go
index 9c9edaa..43f0b08 100644
--- a/common/event_emitter.go
+++ b/common/event_emitter.go
@@ -22,7 +22,6 @@ package common
 
 import (
 	"context"
-	"sync"
 )
 
 // Ensure BaseEventEmitter implements the EventEmitter interface.
@@ -98,10 +97,9 @@ type NavigationEvent struct {
 }
 
 type eventHandler struct {
-	ctx        context.Context
-	ch         chan Event
-	queueMutex *sync.Mutex
-	queue      []Event
+	ctx   context.Context
+	ch    chan Event
+	queue chan Event
 }
 
 // EventEmitter that all event emitters need to implement.
@@ -169,12 +167,14 @@ func (e *BaseEventEmitter) sync(fn func()) {
 
 func (e *BaseEventEmitter) emit(event string, data interface{}) {
 	emitEvent := func(eh eventHandler) {
-		eh.queueMutex.Lock()
-		defer eh.queueMutex.Unlock()
-
+		var e Event
+		select {
+		case e = <-eh.queue:
+		case <-eh.ctx.Done():
+			return
+		}
 		select {
-		case eh.ch <- eh.queue[0]:
-			eh.queue = eh.queue[1:]
+		case eh.ch <- e:
 		case <-eh.ctx.Done():
 			// TODO: handle the error
 		}
@@ -193,12 +193,14 @@ func (e *BaseEventEmitter) emit(event string, data interface{}) {
 				// goroutine responsible for the the first popped element must complete
 				// publishing the event before the next goroutine can pop and work with
 				// the next event for that one handler. Each handler can process events concurrently still.
-
-				handler.queueMutex.Lock()
-				handler.queue = append(handler.queue, Event{typ: event, data: data})
-				handler.queueMutex.Unlock()
-
 				go emitEvent(handler)
+				select {
+				case handler.queue <- Event{typ: event, data: data}:
+				case <-handler.ctx.Done():
+					handlers = append(handlers[:i], handlers[i+1:]...)
+					continue
+				}
+
 				i++
 			}
 		}
@@ -218,7 +220,7 @@ func (e *BaseEventEmitter) on(ctx context.Context, events []string, ch chan Even
 			if !ok {
 				e.handlers[event] = make([]eventHandler, 0)
 			}
-			eh := eventHandler{ctx, ch, &sync.Mutex{}, make([]Event, 0)}
+			eh := eventHandler{ctx, ch, make(chan Event)}
 			e.handlers[event] = append(e.handlers[event], eh)
 		}
 	})
@@ -227,6 +229,6 @@ func (e *BaseEventEmitter) on(ctx context.Context, events []string, ch chan Even
 // OnAll registers a handler for all events.
 func (e *BaseEventEmitter) onAll(ctx context.Context, ch chan Event) {
 	e.sync(func() {
-		e.handlersAll = append(e.handlersAll, eventHandler{ctx, ch, &sync.Mutex{}, make([]Event, 0)})
+		e.handlersAll = append(e.handlersAll, eventHandler{ctx, ch, make(chan Event)})
 	})
 }
```

</details>

Reply:
@inancgumus thanks for the suggestion. Isn't there a chance that the order of the events received by the handler could still be received in the wrong order when using this channel approach?

Situation:
- Handler is emitting an event.
- Emitter is trying to send events to the Handler.
- Since the Handler is busy trying to emit an event, there's a chance that multiple `emitEvent` goroutines could be waiting to send an event to the Handler.
- Which `emitEvent` writes to the channel first? Second? Third? etc.

@imiric I tried to implement your idea (if i understood it correctly) but I was unable to get very far as it's deadlocking. Here's the single commit: https://github.com/grafana/xk6-browser/commit/328dc888d6e7c6ba12487e6fdce289ff6ad6843b

Reply:
@ankur22 If so, shouldn’t there be the same problem with the mutex-queue solution? I mean, the channel solution should work like the mutex-queue solution. What am I missing? :)

Reply:
With queue and mutex: https://go.dev/play/p/rUvviX4qvAw
With chan and no mutex: https://go.dev/play/p/_DN7XRMzeQ9

The key here is the mutex lock, not whether it's a `queue` or a `channel`.

IMO The `queue` is still a better solution since there's a chance that the handler could emit an event while the `emitEvent` could be trying to send an event to the same handler. With an unbuffered queue this would deadlock. With a queue that can grow with the complexity of the test code while avoiding deadlocks.

Reply:
Ah, right, that's because of the lock, see: https://go.dev/play/p/H7m6Rr69tS1